lib/propolis/src/hw/testdev.rs

```diff
@@ -0,0 +1,122 @@
+use std::sync::{
+    atomic::{AtomicUsize, Ordering},
+    Arc,
+};
+
+/// A test device for simulating migration failures.
+pub struct MigrationFailureDevice {
+    log: slog::Logger,
+    exports: AtomicUsize,
+    imports: AtomicUsize,
+    fail: MigrationFailures,
+}
+
+pub struct MigrationFailures {
+    pub exports: usize,
+    pub imports: usize,
+}
+
+#[derive(Clone, Default, Deserialize, Serialize)]
+struct MigrationFailurePayloadV1 {}
+
+impl MigrationFailureDevice {
+    pub const NAME: &'static str = "test-migration-failure";
+
+    pub fn create(log: &slog::Logger, fail: MigrationFailures) -> Arc<Self> {
+        let log =
+            log.new(slog::o!("component" => "testdev", "dev" => Self::NAME));
+        info!(log,
+            "Injecting simulated migration failures";
+            "fail_exports" => %fail.exports,
+            "fail_imports" => %fail.imports,
+        );
+        Arc::new(Self {
+            log,
+            exports: AtomicUsize::new(0),
+            imports: AtomicUsize::new(0),
+            fail,
+        })
+    }
+}
+
+impl Entity for MigrationFailureDevice {
+    fn type_name(&self) -> &'static str {
+        MigrationFailureDevice::NAME
+    }
+    fn migrate(&self) -> Migrator {
+        Migrator::Single(self)
+    }
+}
+
+impl MigrateSingle for MigrationFailureDevice {
+    fn export(
+        &self,
+        _ctx: &MigrateCtx,
+    ) -> Result<PayloadOutput, MigrateStateError> {
+        let export_num = self.exports.fetch_add(1, Ordering::Relaxed);
+        if export_num < self.fail.exports {
+            info!(
+                self.log,
+                "failing export";
+                "export_num" => %export_num,
+                "fail_exports" => %self.fail.exports
+            );
+            return Err(MigrateStateError::Io(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "somebody set up us the bomb",
+            )));
+        }
+
+        info!(
+            self.log,
+            "exporting device";
+            "export_num" => %export_num,
+        );
+        Ok(MigrationFailurePayloadV1 {}.into())
+    }
+
+    fn import(
+        &self,
+        mut offer: PayloadOffer,
+        _ctx: &MigrateCtx,
+    ) -> Result<(), MigrateStateError> {
+        let import_num = self.imports.fetch_add(1, Ordering::Relaxed);
+        let fail = import_num < self.fail.imports;
+        info!(
+            self.log,
+            "exporting device";
+            "import_num" => %import_num,
+            "will_fail" => %fail,
+        );
+        let MigrationFailurePayloadV1 {} = offer.parse()?;
+        if fail {
+            info!(self.log, "failing import");
+            return Err(MigrateStateError::ImportFailed(
+                "you have no chance to survive, make your time".to_string(),
+            ));
+        }
+        Ok(())
+    }
+}
+
+impl Schema<'_> for MigrationFailurePayloadV1 {
+    fn id() -> SchemaId {
+        (MigrationFailureDevice::NAME, 1)
```

Comment:
Since the schema IDs effectively become part of the Committed interface in migration, I'd hesitate to use a reference to the device name here - the latter can be changed on a whim (relatively speaking) whereas changing the former will result in compatibility issues.

Since testing devices are largely an implementation detail in propolis, I'd like to preserve maximum optionality for iteration there.  It'd be nice to make it easy to add or modify devices for test cases without having to worry too much about breaking cross-version tests.